conda_auth/handlers/base.py

```diff
@@ -0,0 +1,158 @@
+        Optionally set a cache to use and configuration parameters to retrieve from
+        ``conda.base.context.context.channel_settings``.
+        """
+        self._context = context
+        self.cache = {} if cache is None else cache
+
+    def get_action_func(self) -> Callable[[str], None]:
+        """Return a callable to be used as the action function for the pre-command plugin hook"""
+
+        def action(command: str):
+            for settings in self._context.channel_settings:
+                if channel := settings.get("channel"):
+                    channel = Channel(channel)
+                    # Only attempt to authenticate for actively used channels
+                    if channel.canonical_name in self._context.channels:
+                        self.authenticate(channel, settings)
+
+        return action
+
+    def authenticate(self, channel: Channel, settings: dict[str, str]) -> None:
+        """Used to retrieve credentials and store them on the ``cache`` property"""
+        if settings.get("auth") == self.get_auth_type():
+            extra_params = {
+                param: settings.get(param) for param in self.get_config_parameters()
+            }
+            self.set_secrets(channel, extra_params)
+
+    @abstractmethod
+    def set_secrets(self, channel: Channel, settings: dict[str, str | None]) -> None:
+        """Implementations should include routine for fetching and storing secrets"""
+
+    @abstractmethod
+    def remove_secrets(self, channel: Channel, settings: dict[str, str | None]) -> None:
+        """Implementations should include routine for removing secrets"""
+
+    @abstractmethod
+    def get_auth_type(self) -> str:
+        """
+        Implementation should return a ``str`` which matches the name of the auth handler type in
+        conda's configuration.
+        """
+
+    @abstractmethod
+    def get_config_parameters(self) -> tuple[str, ...]:
+        """
+        Implementations should return a ``tuple`` of ``str`` which represent the configuration
+        values to use in the ``context.channel_settings`` object.
+        """
+
+    @abstractmethod
+    def get_keyring_id(self, channel_name: str) -> str:
+        """
+        Implementation should return the keyring id that will be used by the manager classes
+        """
+
+
+class CacheChannelAuthBase(ChannelAuthBase):
+    """
+    Adds a class instance cache object for storage of authentication information.
+    """
+
+    def __init__(self, channel_name: str):
+        """
+        Makes sure we have initialized the cache object.
+        """
+        super().__init__(channel_name)
+
+        if not hasattr(self, "_cache"):
+            raise CondaAuthError(
+                "Cache not initialized on class; please run `BasicAuthHandler.set_cache`"
+                " before using"
+            )
+
+    @classmethod
+    def set_cache(cls, cache: dict[str, Any]) -> None:
+        cls._cache = cache
+
+
+def test_credentials(func):
```

Comment:
To not trip up pytest given the funciton name, I would rename that to something like
```suggestion
def verified_credentials(func):
```